File: Neetcode150/C#/10.Tries.cs
using System;
using System.Collections.Generic;

public class TrieNode
{
    public Dictionary<char, TrieNode> Children { get; } = new Dictionary<char, TrieNode>();
    public bool IsEnd { get; set; }
    public string Word { get; set; } // For Word Search II
}

public class Trie
{
    private readonly TrieNode root;

    public Trie()
    {
        root = new TrieNode();
    }

    public void Insert(string word)
    {
        var node = root;
        foreach (var ch in word)
        {
            if (!node.Children.ContainsKey(ch))
                node.Children[ch] = new TrieNode();
            node = node.Children[ch];
        }
        node.IsEnd = true;
    }

    public bool Search(string word)
    {
        var node = root;
        foreach (var ch in word)
        {
            if (!node.Children.ContainsKey(ch))
                return false;
            node = node.Children[ch];
        }
        return node.IsEnd;
    }

    public bool StartsWith(string prefix)
    {
        var node = root;
        foreach (var ch in prefix)
        {
            if (!node.Children.ContainsKey(ch))
                return false;
            node = node.Children[ch];
        }
        return true;
    }
}

public class WordDictionary
{
    private readonly TrieNode root;

    public WordDictionary()
    {
        root = new TrieNode();
    }

    public void AddWord(string word)
    {
        var node = root;
        foreach (var ch in word)
        {
            if (!node.Children.ContainsKey(ch))
                node.Children[ch] = new TrieNode();
            node = node.Children[ch];
        }
        node.IsEnd = true;
    }

    public bool Search(string word)
    {
        return Dfs(root, word, 0);
    }

    private bool Dfs(TrieNode node, string word, int index)
    {
        if (index == word.Length)
            return node.IsEnd;

        char ch = word[index];
        if (ch == '.')
        {
            foreach (var child in node.Children.Values)
            {
                if (Dfs(child, word, index + 1))
                    return true;
            }
            return false;
        }
        else
        {
            if (!node.Children.ContainsKey(ch))
                return false;
            return Dfs(node.Children[ch], word, index + 1);
        }
    }
}

public static class WordSearchII
{
    public static IList<string> FindWords(char[][] board, string[] words)
    {
        var root = new TrieNode();
        foreach (var word in words)
        {
            var node = root;
            foreach (var ch in word)
            {
                if (!node.Children.ContainsKey(ch))
                    node.Children[ch] = new TrieNode();
                node = node.Children[ch];
            }
            node.IsEnd = true;
            node.Word = word;
        }

        var result = new List<string>();
        int rows = board.Length;
        int cols = board[0].Length;

        void Dfs(int r, int c, TrieNode node)
        {
            if (r < 0 || c < 0 || r >= rows || c >= cols)
                return;

            char ch = board[r][c];
            if (ch == '#' || !node.Children.ContainsKey(ch))
                return;

            node = node.Children[ch];
            if (node.IsEnd)
            {
                result.Add(node.Word);
                node.IsEnd = false; // avoid duplicates
            }

            board[r][c] = '#'; // mark visited
            Dfs(r + 1, c, node);
            Dfs(r - 1, c, node);
            Dfs(r, c + 1, node);
            Dfs(r, c - 1, node);
            board[r][c] = ch; // restore
        }

        for (int r = 0; r < rows; r++)
        {
            for (int c = 0; c < cols; c++)
            {
                Dfs(r, c, root);
            }
        }

        return result;
    }
}

public static class Program
{
    static void Main()
    {
        var trie = new Trie();
        trie.Insert("apple");
        Console.WriteLine(trie.Search("apple")); // true
        Console.WriteLine(trie.Search("app")); // false
        Console.WriteLine(trie.StartsWith("app")); // true
        trie.Insert("app");
        Console.WriteLine(trie.Search("app")); // true

        Console.WriteLine("------------------------------------------------------------");

        var wordDict = new WordDictionary();
        wordDict.AddWord("bad");
        wordDict.AddWord("dad");
        wordDict.AddWord("mad");
        Console.WriteLine(wordDict.Search("pad")); // false
        Console.WriteLine(wordDict.Search("bad")); // true
        Console.WriteLine(wordDict.Search(".ad")); // true
        Console.WriteLine(wordDict.Search("b..")); // true    

        Console.WriteLine("------------------------------------------------------------");

        var board = new Char[][] {
            new Char[] {'o','a','a','n'},
            new Char[] {'e','t','a','e'},
            new Char[] {'i','h','k','r'},
            new Char[] {'i','f','l','v'}
        };
        var words = new String[] { "oath", "pea", "eat", "rain" };
        var foundWords = WordSearchII.FindWords(board, words);
        foreach (var w in foundWords)
            Console.WriteLine(w);
    }
}

// Example usage:
// var trie = new Trie();
// trie.Insert("apple");
// Console.WriteLine(trie.Search("apple")); // true
// Console.WriteLine(trie.Search("app")); // false
// Console.WriteLine(trie.StartsWith("app")); // true
// trie.Insert("app");
// Console.WriteLine(trie.Search("app")); // true

// var wordDict = new WordDictionary();
// wordDict.AddWord("bad");
// wordDict.AddWord("dad");
// wordDict.AddWord("mad");
// Console.WriteLine(wordDict.Search("pad")); // false
// Console.WriteLine(wordDict.Search("bad")); // true
// Console.WriteLine(wordDict.Search(".ad")); // true
// Console.WriteLine(wordDict.Search("b..")); // true

// char[][] board = new char[][] {
//     new char[] {'o','a','a','n'},
//     new char[] {'e','t','a','e'},
//     new char[] {'i','h','k','r'},
//     new char[] {'i','f','l','v'}
// };
// string[] words = new string[] {"oath","pea","eat","rain"};
// var foundWords = WordSearchII.FindWords(board, words);
// foreach (var w in foundWords)
//     Console.WriteLine(w);

File: Neetcode150/10.Tries.js
// Implement Trie (Prefix Tree)
// Each node represents a character. insert builds paths, search checks full word existence, startsWith checks prefix existence

class TrieNode {
    constructor() {
        this.children = {};
        this.isEnd = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    insert(word) {
        let node = this.root;
        for (let char of word) {
            if (!node.children[char]) node.children[char] = new TrieNode();
            node = node.children[char];
        }
        node.isEnd = true;
    }

    search(word) {
        let node = this.root;
        for (let char of word) {
            if (!node.children[char]) return false;
            node = node.children[char];
        }
        return node.isEnd;
    }

    startsWith(prefix) {
        let node = this.root;
        for (let char of prefix) {
            if (!node.children[char]) return false;
            node = node.children[char];
        }
        return true;
    }
}

const trie = new Trie();
trie.insert("apple");
console.log(trie.search("apple")); // true
console.log(trie.search("app")); // false
console.log(trie.startsWith("app")); // true
trie.insert("app");
console.log(trie.search("app")); // true


// Design Add and Search Words Data Structure
// Similar to Trie, but supports wildcard ".". Use DFS to explore all possible children when encountering ".".
// Trie + DFS for wildcard
class WordDictionary {
    constructor() {
        this.root = new TrieNode();
    }

    addWord(word) {
        let node = this.root;
        for (let char of word) {
            if (!node.children[char]) node.children[char] = new TrieNode();
            node = node.children[char];
        }
        node.isEnd = true;
    }

    search(word) {
        function dfs(node, i) {
            if (i === word.length) return node.isEnd;
            let char = word[i];
            if (char === ".") {
                for (let child in node.children) {
                    if (dfs(node.children[child], i + 1)) return true;
                }
                return false;
            } else {
                if (!node.children[char]) return false;
                return dfs(node.children[char], i + 1);
            }
        }
        return dfs(this.root, 0);
    }
}
const wordDict = new WordDictionary();
wordDict.addWord("bad");
wordDict.addWord("dad");
wordDict.addWord("mad");
console.log(wordDict.search("pad")); // false
console.log(wordDict.search("bad")); // true
console.log(wordDict.search(".ad")); // true
console.log(wordDict.search("b..")); // true    

// Word Search II
// Build a Trie of words. Then DFS on the board, following Trie paths. Mark visited cells with "#". Collect words when reaching isEnd.
// Trie + DFS backtracking on grid
function findWords(board, words) {
    let root = new TrieNode();
    for (let word of words) {
        let node = root;
        for (let char of word) {
            if (!node.children[char]) node.children[char] = new TrieNode();
            node = node.children[char];
        }
        node.isEnd = true;
        node.word = word; // store word at end node
    }

    let result = [];
    let rows = board.length, cols = board[0].length;

    function dfs(r, c, node) {
        if (r < 0 || c < 0 || r >= rows || c >= cols) return;
        let char = board[r][c];
        if (char === "#" || !node.children[char]) return;

        node = node.children[char];
        if (node.isEnd) {
            result.push(node.word);
            node.isEnd = false; // avoid duplicates
        }

        board[r][c] = "#"; // mark visited
        dfs(r + 1, c, node);
        dfs(r - 1, c, node);
        dfs(r, c + 1, node);
        dfs(r, c - 1, node);
        board[r][c] = char; // restore
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            dfs(r, c, root);
        }
    }

    return result;
}
console.log(findWords([["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]], ["oath","pea","eat","rain"])); // ["oath", "eat"]
